File: automedal/agent/tools/fs.py
"""Filesystem tools — read_file, write_file, edit_file, list_dir, grep.

All paths are resolved relative to REPO_ROOT and rejected if they escape it.
Tool bodies are ported from the prior automedal/agent_runtime.py so existing
prompts (which were written against pi's identical surface) port unchanged.
"""

from __future__ import annotations

import fnmatch
import re
from pathlib import Path
from typing import Iterable

from automedal.agent.tools.base import REPO_ROOT, Tool, ToolResult, _safe, make_tool


# ── implementations (sync — kernel wraps in a coroutine) ─────────────────────

def _read_file(path: str) -> ToolResult:
    return ToolResult(_safe(path).read_text(encoding="utf-8"))


def _write_file(path: str, content: str) -> ToolResult:
    p = _safe(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(content, encoding="utf-8")
    return ToolResult(f"wrote {path} ({len(content)} chars)")


def _edit_file(path: str, old: str, new: str) -> ToolResult:
    p = _safe(path)
    txt = p.read_text(encoding="utf-8")
    n = txt.count(old)
    if n != 1:
        return ToolResult(f"error: old string appears {n} times (needs exactly 1)", ok=False)
    p.write_text(txt.replace(old, new), encoding="utf-8")
    return ToolResult(f"edited {path}")


def _list_dir(path: str = ".") -> ToolResult:
    p = _safe(path)
    if not p.is_dir():
        return ToolResult(f"error: {path} is not a directory", ok=False)
    entries = sorted(p.iterdir(), key=lambda x: (not x.is_dir(), x.name.lower()))
    return ToolResult("\n".join(f"{'d' if e.is_dir() else 'f'}  {e.name}" for e in entries))


def _grep(pattern: str, path: str = ".", glob: str = "*") -> ToolResult:
    root = _safe(path)
    rx = re.compile(pattern)
    hits: list[str] = []
    walker: Iterable[Path]
    walker = [root] if root.is_file() else root.rglob("*")
    for f in walker:
        if not f.is_file() or not fnmatch.fnmatch(f.name, glob):
            continue
        try:
            for lineno, line in enumerate(f.read_text(encoding="utf-8").splitlines(), 1):
                if rx.search(line):
                    rel = f.relative_to(REPO_ROOT)
                    hits.append(f"{rel}:{lineno}: {line.rstrip()}")
                    if len(hits) >= 80:
                        return ToolResult("\n".join(hits) + "\n... (truncated at 80 matches)")
        except (UnicodeDecodeError, PermissionError):
            continue
    return ToolResult("\n".join(hits) if hits else "(no matches)")


# ── Tool objects ─────────────────────────────────────────────────────────────

READ_FILE = make_tool(
    name="read_file",
    description="Read a UTF-8 file. Path is resolved relative to the repo root.",
    schema={
        "type": "object",
        "properties": {"path": {"type": "string", "description": "Relative path"}},
        "required": ["path"],
    },
    fn=_read_file,
)

WRITE_FILE = make_tool(
    name="write_file",
    description="Create or overwrite a UTF-8 file relative to the repo root.",
    schema={
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "content": {"type": "string"},
        },
        "required": ["path", "content"],
    },
    fn=_write_file,
)

EDIT_FILE = make_tool(
    name="edit_file",
    description=(
        "Replace `old` with `new` in `path`. `old` must appear EXACTLY once "
        "or the call fails."
    ),
    schema={
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "old":  {"type": "string"},
            "new":  {"type": "string"},
        },
        "required": ["path", "old", "new"],
    },
    fn=_edit_file,
)

LIST_DIR = make_tool(
    name="list_dir",
    description="List immediate children of a directory relative to the repo root.",
    schema={
        "type": "object",
        "properties": {"path": {"type": "string", "default": "."}},
        "required": [],
    },
    fn=_list_dir,
)

GREP = make_tool(
    name="grep",
    description=(
        "Search for `pattern` (regex) in files matching `glob` under `path`. "
        "Returns up to 80 'relpath:lineno: line' matches."
    ),
    schema={
        "type": "object",
        "properties": {
            "pattern": {"type": "string"},
            "path":    {"type": "string", "default": "."},
            "glob":    {"type": "string", "default": "*"},
        },
        "required": ["pattern"],
    },
    fn=_grep,
)


FS_TOOLS: list[Tool] = [READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP]

File: internal/agent/tools/fs.go
// Filesystem tools — read_file, write_file, edit_file, list_dir, grep.
//
// All paths are resolved relative to RepoRoot and rejected if they escape it.
// Tool surface mirrors automedal/agent/tools/fs.py.
package tools

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// ── implementations ──────────────────────────────────────────────────

func runReadFile(ctx context.Context, args map[string]any) (ToolResult, error) {
	path := StrArg(args, "path", "")
	if path == "" {
		return MissingArg("path"), nil
	}
	p, err := Safe(path)
	if err != nil {
		return ToolResult{}, err
	}
	b, err := os.ReadFile(p)
	if err != nil {
		return Error("error: %s", err.Error()), nil
	}
	return Result(string(b)), nil
}

func runWriteFile(ctx context.Context, args map[string]any) (ToolResult, error) {
	path := StrArg(args, "path", "")
	if path == "" {
		return MissingArg("path"), nil
	}
	content := StrArg(args, "content", "")
	p, err := Safe(path)
	if err != nil {
		return ToolResult{}, err
	}
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return Error("error: %s", err.Error()), nil
	}
	if err := os.WriteFile(p, []byte(content), 0o644); err != nil {
		return Error("error: %s", err.Error()), nil
	}
	return Result(fmt.Sprintf("wrote %s (%d chars)", path, len(content))), nil
}

func runEditFile(ctx context.Context, args map[string]any) (ToolResult, error) {
	path := StrArg(args, "path", "")
	if path == "" {
		return MissingArg("path"), nil
	}
	old := StrArg(args, "old", "")
	if old == "" {
		return MissingArg("old"), nil
	}
	newStr := StrArg(args, "new", "")
	p, err := Safe(path)
	if err != nil {
		return ToolResult{}, err
	}
	b, err := os.ReadFile(p)
	if err != nil {
		return Error("error: %s", err.Error()), nil
	}
	txt := string(b)
	n := strings.Count(txt, old)
	if n != 1 {
		return Error("error: old string appears %d times (needs exactly 1)", n), nil
	}
	if err := os.WriteFile(p, []byte(strings.Replace(txt, old, newStr, 1)), 0o644); err != nil {
		return Error("error: %s", err.Error()), nil
	}
	return Result(fmt.Sprintf("edited %s", path)), nil
}

func runListDir(ctx context.Context, args map[string]any) (ToolResult, error) {
	path := StrArg(args, "path", ".")
	p, err := Safe(path)
	if err != nil {
		return ToolResult{}, err
	}
	info, err := os.Stat(p)
	if err != nil {
		return Error("error: %s", err.Error()), nil
	}
	if !info.IsDir() {
		return Error("error: %s is not a directory", path), nil
	}
	entries, err := os.ReadDir(p)
	if err != nil {
		return Error("error: %s", err.Error()), nil
	}
	sort.Slice(entries, func(i, j int) bool {
		di, dj := entries[i].IsDir(), entries[j].IsDir()
		if di != dj {
			return di // dirs first
		}
		return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
	})
	var lines []string
	for _, e := range entries {
		mark := "f"
		if e.IsDir() {
			mark = "d"
		}
		lines = append(lines, fmt.Sprintf("%s  %s", mark, e.Name()))
	}
	return Result(strings.Join(lines, "\n")), nil
}

func runGrep(ctx context.Context, args map[string]any) (ToolResult, error) {
	pattern := StrArg(args, "pattern", "")
	if pattern == "" {
		return MissingArg("pattern"), nil
	}
	rx, err := regexp.Compile(pattern)
	if err != nil {
		return Error("error: bad regex: %s", err.Error()), nil
	}
	path := StrArg(args, "path", ".")
	glob := StrArg(args, "glob", "*")
	root, err := Safe(path)
	if err != nil {
		return ToolResult{}, err
	}
	repoRoot := RepoRoot()

	hits := make([]string, 0, 80)
	walkOne := func(f string) (stop bool) {
		base := filepath.Base(f)
		if matched, _ := filepath.Match(glob, base); !matched {
			return false
		}
		b, err := os.ReadFile(f)
		if err != nil {
			return false
		}
		if !isText(b) {
			return false
		}
		rel, _ := filepath.Rel(repoRoot, f)
		for i, line := range strings.Split(string(b), "\n") {
			if rx.MatchString(line) {
				hits = append(hits, fmt.Sprintf("%s:%d: %s", rel, i+1, strings.TrimRight(line, "\r ")))
				if len(hits) >= 80 {
					return true
				}
			}
		}
		return false
	}

	info, err := os.Stat(root)
	if err != nil {
		return Error("error: %s", err.Error()), nil
	}
	if !info.IsDir() {
		walkOne(root)
	} else {
		filepath.Walk(root, func(p string, fi os.FileInfo, err error) error {
			if err != nil || fi.IsDir() {
				return nil
			}
			if walkOne(p) {
				return filepath.SkipAll
			}
			return nil
		})
	}
	if len(hits) == 0 {
		return Result("(no matches)"), nil
	}
	out := strings.Join(hits, "\n")
	if len(hits) >= 80 {
		out += "\n... (truncated at 80 matches)"
	}
	return Result(out), nil
}

// isText is a small heuristic: if the first 1KB has no NULL bytes and is
// valid UTF-8 enough we treat it as text.
func isText(b []byte) bool {
	n := len(b)
	if n > 1024 {
		n = 1024
	}
	for i := 0; i < n; i++ {
		if b[i] == 0 {
			return false
		}
	}
	return true
}

// ── Tool definitions ─────────────────────────────────────────────────

var ReadFile = Tool{
	Name:        "read_file",
	Description: "Read a UTF-8 file. Path is resolved relative to the repo root.",
	Schema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path": map[string]any{"type": "string", "description": "Relative path"},
		},
		"required": []string{"path"},
	},
	Run: runReadFile,
}

var WriteFile = Tool{
	Name:        "write_file",
	Description: "Create or overwrite a UTF-8 file relative to the repo root.",
	Schema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path":    map[string]any{"type": "string"},
			"content": map[string]any{"type": "string"},
		},
		"required": []string{"path", "content"},
	},
	Run: runWriteFile,
}

var EditFile = Tool{
	Name: "edit_file",
	Description: "Replace `old` with `new` in `path`. `old` must appear EXACTLY once " +
		"or the call fails.",
	Schema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path": map[string]any{"type": "string"},
			"old":  map[string]any{"type": "string"},
			"new":  map[string]any{"type": "string"},
		},
		"required": []string{"path", "old", "new"},
	},
	Run: runEditFile,
}

var ListDir = Tool{
	Name:        "list_dir",
	Description: "List immediate children of a directory relative to the repo root.",
	Schema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path": map[string]any{"type": "string", "default": "."},
		},
		"required": []string{},
	},
	Run: runListDir,
}

var Grep = Tool{
	Name: "grep",
	Description: "Search for `pattern` (regex) in files matching `glob` under `path`. " +
		"Returns up to 80 'relpath:lineno: line' matches.",
	Schema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"pattern": map[string]any{"type": "string"},
			"path":    map[string]any{"type": "string", "default": "."},
			"glob":    map[string]any{"type": "string", "default": "*"},
		},
		"required": []string{"pattern"},
	},
	Run: runGrep,
}

// FSTools is the canonical filesystem tool bundle, in the order phases use.
var FSTools = []Tool{ReadFile, WriteFile, EditFile, ListDir, Grep}
